core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
```

Comment:
This two block separation suggest that this could be split into two interfaces...

Reply:
My suggestion would be this:
1) Get this interface nailed down how we like it
2) Introduce a parallel interface for reading lines of text
3) Refactor the seekable part into a standalone interface that is used in both

All of which would be done _long_ before even an alpha release.  I just find it easier to keep it in one interface while we work things out.

Reply:
Even in that case, when implementations start to get more complicated it will be difficult to uncouple the two of them. But I don't know what is the timeframe for splitting them that you are suggesting.